### DISCUSSION TITLE
Is it a good idea to store an array of previous states in the machine `context`?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [statelyai/xstate#3019] [Q&A]
Note: I have never used state machines before and reactive programming is all new to me. I hope you’ll understand that. :wink:

---

Besides monitoring the current state, I have a requirement to (let’s say) cache the previous states up until a specified period (e.g. past hour or past 16 hours). According to my measurements, there could be maximally about 600 hundred such states in 16 hours in my use case.

I was thinking about creating `context: {currentState: {}, previousStates: []}` and update it every time the state changes.

I read about [XState’s history state](https://xstate.js.org/docs/guides/history.html#history-state-configuration), however, IMHO it is not useful to me, as I need to get all the previous states at once (i.e. get the entire `context.previousStates` array).

I would use this as a cache only; to persist the states I use a database.

I am open to any suggestions how I could use XState for this.

---

In order to be more specific in the description of my use case: I would like to create a state machine:
- for the current state (`{created: Date, state: StateEnum}`) without need for any historical values;
- for the state intervals (periods without changing the state; e.g. `{startTime: Date, endTime: Date, state: StateEnum}`):
  - in here I need:
    - the current interval;
    - an array of all intervals in the past hour (modifying the oldest interval start time);
    - an array of all intervals in any workshift of the current day (e.g. `morningShift`, `afternoonShift`, `nightShift`), reset at the start time of the first shift in a day (e.g. `morningShift`).

---

**Update:**

Actually, I have trouble with finding a way to gather the current state data every second. I could use `cron`, but I think this should be doable using `xstate` and `setInterval()`.

I need to gather the data and process it every second. When the `context` is updated, another machine (with the state intervals) would process the current state. At least this is how I look at this.

I’ve found [`Scheduler`](https://xstate.js.org/api/classes/scheduler.html) class, but I haven’t found an example on how to use it.

I’ve also read about [invoking callbacks](https://xstate.js.org/docs/guides/communication.html#invoking-callbacks), but it doesn’t seem right to clear the interval. Maybe it is actually good.

Another way would be to not creating `states` as the states of the CNC machines, but as states of the CNC state data processing, i.e. `GET_DATA`, `TIME_OUT` which both would trigger an action and change the state to the other.

---

**Update 2:** Would [the Timer example](https://xstate.js.org/docs/tutorials/7guis/timer.html#modeling) be the proper way to implement my use case? I uses operation states instead of CNC machine states, but it works. I’d just remove the paused state.

---

**Update 3:**

Is this the right way to do?

```js
const {assign, createMachine, interpret} = require('xstate')

const stateMachine = createMachine({
  initial: 'timer',
  context: {
    created: new Date(),
    state: 100000
  },
  states: {
    timer: {
      invoke: {
        src: context => cb => {
          const interval = setInterval(() => {
            cb('GET_STATE')
          }, 1000)

          return () => {
            clearInterval(interval)
          }
        }
      },
      on: {
        GET_STATE: {
          actions: assign({
            created: context => new Date(),
            // Note: This is just for example
            state: context => context.state - 1
          })
        }
      }
    }
  }
})

const stateService = interpret(stateMachine)
  .onTransition((state) => console.log(`context : ${JSON.stringify(state.context, null, 2)}`))
  .start()
```

**Comment by tukusejssirs** [comment]
**Update 3:**

I changed the code substantially (see [here](https://stately.ai/viz/2c965f17-f447-4bbe-8469-4cd7072adc46#) the current code), however, I have troubles with sending the `context` of one machine to the invoked machine. I also opened a question regarding this on [SO](https://stackoverflow.com/questions/71124662/how-to-invoke-a-machine-with-context-of-the-machine-that-invokes-that-machine).

**Reply by Andarist** [reply]
`data` is only used when initializing the child machine. With the given solution you could implement a pattern that would resemble a request. Send a request for the current state to the parent machine and just respond to the caller with the state. You can use our `respond` action creator which should make it slightly easier for you.

**Reply by Andarist** [reply]
I see that you are already, sort of, implementing that in your Update 3 but your latest code doesn't use that. I wonder what was the problem with that code there? Were some requirements not satisfied by it?

**Reply by tukusejssirs** [reply]
Thanks, @Andarist, for trying to help me out! :pray:

Could you please pinpoint the issue in the [latest code](https://stately.ai/viz/2c965f17-f447-4bbe-8469-4cd7072adc46#) (same code/link as in the update 3 I posted above)?

All I need is to use the context of the _invoking_ machine as sort of a parameter a in the `invoke` (let’s say) function and thus make it available in the _invoked_ machine. Is it possible?

> I wonder what was the problem with that code there? Were some requirements not satisfied by it?

Currently, when I send an event (e.g. `GREEN`), the state of `currentStateMachine` changes as expected, but it seems like the `context` is not ‘sent’ the invoked machine (`intervalsMachine`), as the `context` of the `intervalsMachine` changes to `{state: undefined}`.

**Reply by Andarist** [reply]
Your `intervalsMachine` listens for `NEW_STATE` event but such an event is not sent to it anywhere. I would expect the `currentStateMachine` to send it to the `intervalsMachine`

> All I need is to use the context of the invoking machine as sort of a parameter a in the invoke (let’s say) function and thus make it available in the invoked machine. Is it possible?

That's currently not possible. You can only pass the context of the invoking machine to the invoked one when initializing the invoked one but it would be the "frozen" context. I mean, the context in XState is treated as an immutable container and we do not make mutations to it. So if you share that object with the invoked machine then it won't see any updates being made to it because the updates will be done to the newly created copy of the context in the parent machine.

> Currently, when I send an event (e.g. GREEN), the state of currentStateMachine changes as expected, but it seems like the context is not ‘sent’ the invoked machine (intervalsMachine), as the context of the intervalsMachine changes to {state: undefined}.

You need to send an event explicitly to the `intervalsMachine`. It won't sync automatically.

**Reply by tukusejssirs** [reply]
> That's currently not possible.

Okay. As I understand it, I would need to use another variable for the current state (`{created: Date, state: number}`) instead of `context`, right? Or do you think that is a better way to do what I want to do?

> You need to send an event explicitly to the `intervalsMachine`. It won't sync automatically.

You mean something like `currentService.send({type: 'GREEN'})`? Therefore the entire transport should be outside of the state machine.

What do you think: has usage of the state machines (XState) any benefit in my use case?

**Reply by Andarist** [reply]
It could look something like this: https://stately.ai/viz/f5f707b8-189f-4638-b976-29df5ed931da . Note that I've made changes very quickly and u might want to refactor this to your liking.

> What do you think: has usage of the state machines (XState) any benefit in my use case?

It's hard to tell without knowing more about your use case but I'd say that if your real system is more complicated than this then it might be beneficial for you to use state machines (it usually is for business oriented logic)

**Reply by tukusejssirs** [reply]
> It's hard to tell without knowing more about your use case but I'd say that if your real system is more complicated than this then it might be beneficial for you to use state machines (it usually is for business oriented logic)

Well, basically I want to do three things:
- get current state every second;
- calculate state intervals from the current state:
   - after that, I need to persist them into the DB;
   - I also need to keep all intervals in the past hour in memory;
- calculate cumulative data per predefined periods (hour, shift, all shifts in a day, day, week, month, year):
   - I need to persist them into the DB;
   - I also need to keep them all in memory.

This is only one feature scope, but there would be 4-5 difference scopes with similar logic per CNC machine, and there are 60 CNC machines.

---

I looked at the code you created for me. I see that the `intervalsMachine` (child state machine) needs to request the data from `currentStateMachine` (parent state machine). Thank you very much. I need to go through it a bit more though. :wink:

One question about `after`: the machine simply times out for the specified period _after_ it enters the `waiting` state, right? Therefore intervals processing occurs every `after` timout value + the time needed to process the requested data, right?

**Reply by Andarist** [reply]
> One question about after: the machine simply times out for the specified period after it enters the waiting state, right?

Yes, note though that if you exit this state then the timer gets canceled.

> Therefore intervals processing occurs every after timout value + the time needed to process the requested data, right?

Yes, but I would say that "process the requested data" is here considered to be almost 0 time. Unless your processing is async - but then you need to model this explicitly, handle errors, etc.

**Reply by tukusejssirs** [reply]
> Yes, but I would say that "process the requested data" is here considered to be almost 0 time. Unless your processing is async - but then you need to model this explicitly, handle errors, etc.

Yeah, in the overly simplified example, the execution time is almost zero, however, using the actual code, it takes 200-500 milliseconds, not including the time to get the current data.

Is there a way to create a queue of awaited current states if it would take longer to process than a second (including the time required to get the current data)? Also, it would be required that the processing of the current data by the `intervalsMachine` would not stop the `currentMachine` from working. I believe this cannot by done within the state machines and it needs to be done outside of them. Any idea is welcome! :pray:

**Reply by Andarist** [reply]
> Yeah, in the overly simplified example, the execution time is almost zero, however, using the actual code, it takes 200-500 milliseconds, not including the time to get the current data.

That sounds like a lot if all of that happens in a synchronous function. I would probably recommend trying to optimize this - unless this is just a script that doesn't need to handle any parallel work. That goes beyond the subject of this thread though 😉 

> Is there a way to create a queue of awaited current states if it would take longer to process than a second (including the time required to get the current data)? Also, it would be required that the processing of the current data by the intervalsMachine would not stop the currentMachine from working. I believe this cannot by done within the state machines and it needs to be done outside of them. Any idea is welcome! 🙏

Freeing up computation resources is up to you here - it doesn't quite matter if you are using state machines here or any different technique. In JavaScript, If you are blocking the main thread then no other work can be done at the same time (on this thread). From the given description I assume that this particular problem is more with your code than with how you are structuring this in a machine etc. You need to make your processing async or optimize it a lot.

**Reply by tukusejssirs** [reply]
> That sounds like a lot if all of that happens in a synchronous function.

Currently (without state machines), I created a cronjob (using `cron` package) to get the data once a second and create the intervals (as well as some other stuff) and the cronjob callback is an asynchronous function. It _mostly_ works, occasionally it fails to process the data because the processing of a single state takes a bit longer than a second. That’s way I think I might need to create a state queue to process the data, as the data needs to be processed chronologically.

> I would probably recommend trying to optimize this - unless this is just a script that doesn't need to handle any parallel work. That goes beyond the subject of this thread though :wink:

I get it is out of scope of this GitHub Discussion and XState. Just for completeness, here’s the function I currently use to calculate the intervals from individual states. It is quite complex, but I need to take into account the time difference (if it is more than `maxIntDiff`, i.e. four seconds, it is considered to be an error, unless the CNC machine is turned off; only then the states are compared), but we need to follow the rules. There might be a better algorithm though.

<details>
  <summary>Function to calculate intervals from individual states</summary>

```typescript
private async calculateStatusInterval(
  {machine, status}: {machine: number; status: CurrentLightIndicatorStatusData}
): Promise<{insert?: CurrentLightIndicatorIntervalData[], update?: CurrentLightIndicatorIntervalData[]}> {
  const schema = `m${machine}`
  const changes: {insert?: CurrentLightIndicatorIntervalData[], update?: CurrentLightIndicatorIntervalData[]} = {}

  if (this.currentLightIndicatorData[schema].intervals.hasOwnProperty('latest') && Object.keys(this.currentLightIndicatorData[schema].intervals.latest).length === 0) {
    // There are no intervals yet

    changes.insert = []

    // Insert `isMonitoringOff = true` from start time of the first workshift until the start time of the first interval
    if (status.created > this.shiftTimes.startFirstShift.valueOf()) {
      changes.insert.push({
        startTime: this.shiftTimes.startFirstShift.valueOf(),
        length: status.created - this.shiftTimes.startFirstShift.valueOf(),
        status: StateEnum.IS_MONITORING_OFF
      })
    }

    // Insert an interval with the current status
    changes.insert.push({
      startTime: status.created,
      length: 0,
      status: status.status
    })
  } else if (Object.keys(this.currentLightIndicatorData[schema].intervals.latest).length > 0) {
    if (!this.currentLightIndicatorData[schema].intervals.hasOwnProperty('latestChanges') && Object.keys(this.currentLightIndicatorData[schema].intervals.latestChanges).length > 0) {
      // There is at least one interval in the DB, but none is cached yet

      changes.insert = []

      // Create `isMonitoringOff = true` interval from end time of the previous interval to `status.created` right after the monitoring was enabled
      changes.insert.push({
        startTime: this.currentLightIndicatorData[schema].intervals.latest.startTime + this.currentLightIndicatorData[schema].intervals.latest.length,
        length:
          status.created - this.currentLightIndicatorData[schema].intervals.latest.startTime - this.currentLightIndicatorData[schema].intervals.latest.length,
        status: StateEnum.IS_MONITORING_OFF
      })

      // Insert an interval with the current status
      changes.insert.push({
        startTime: status.created,
        length: 0,
        status: status.status
      })
    } else {
      // There is at least one cached interval

      // Compare dates
      const dateDiff = status.created - this.currentLightIndicatorData[schema].intervals.latest.startTime - this.currentLightIndicatorData[schema].intervals.latest.length

      // FIXME: This is a workaround. It simply skips any state records when a following record was already processed.
      if (dateDiff < 0) {
        this.logger.log(`[$schema] Skipping processing the following state data because the date diff between end time of the previous interval (${new Date(this.currentLightIndicatorData[schema].intervals.latest.startTime + this.currentLightIndicatorData[schema].intervals.latest.length).toISOString()}) and the current timestamp (${new Date(status.created).toISOString()}) is negative (${dateDiff}) : ` + JSON.stringify(status, null, 2))
        return this.currentLightIndicatorData[schema].intervals.latestChanges
      }

      if (status.status === this.currentLightIndicatorData[schema].intervals.latest.status && (
        dateDiff >= 0 && dateDiff <= maxIntDiff
        || status.status === StateEnum.IS_TURNED_OFF && this.currentLightIndicatorData[schema].intervals.latest.status === StateEnum.IS_TURNED_OFF
        || status.status === StateEnum.ERROR && this.currentLightIndicatorData[schema].intervals.latest.status === StateEnum.ERROR
        || status.status === StateEnum.IS_MONITORING_OFF && this.currentLightIndicatorData[schema].intervals.latest.status === StateEnum.IS_MONITORING_OFF
      )) {
        // Update the previous interval
        changes.update = []
        changes.update.push(this.currentLightIndicatorData[schema].intervals.latest)
        changes.update[changes.update.length - 1].length = status.created - this.currentLightIndicatorData[schema].intervals.latest.startTime
      } else {
        // The status in previous interval and current status are not the same OR if `dateDiff > maxIntDiff`

        if (dateDiff >= 0 && dateDiff <= maxIntDiff) {
          // Update the previous interval
          changes.update = []
          changes.update.push(this.currentLightIndicatorData[schema].intervals.latest)
          changes.update[changes.update.length - 1].length = status.created - this.currentLightIndicatorData[schema].intervals.latest.startTime
        } else {
          // `dateDiff > maxIntDiff`

          changes.insert = []

          // Create additional interval
          changes.insert.push({
            startTime: this.currentLightIndicatorData[schema].intervals.latest.startTime + this.currentLightIndicatorData[schema].intervals.latest.length,
            length:
              status.created - this.currentLightIndicatorData[schema].intervals.latest.startTime - this.currentLightIndicatorData[schema].intervals.latest.length,
            status: undefined
          })

          if (status.status === StateEnum.IS_TURNED_OFF) {
            // Create `isTurnedOff = true` interval
            changes.insert[changes.insert.length - 1].status = StateEnum.IS_TURNED_OFF
          } else {
            // Create `error = true` interval
            changes.insert[changes.insert.length - 1].status = StateEnum.ERROR
          }
        }

        // Create a new status interval unless current status is `isTurnedOff = true` or `error = true`
        if (status.status !== StateEnum.IS_TURNED_OFF && status.status !== StateEnum.ERROR) {
          changes.insert = changes.insert ?? []
          changes.insert.push({
            startTime: status.created,
            length: 0,
            status: status.status
          })
        }
      }
    }
  }

  return changes
}
```

</details>

> Freeing up computation resources is up to you here - it doesn't quite matter if you are using state machines here or any different technique. In JavaScript, If you are blocking the main thread then no other work can be done at the same time (on this thread). From the given description I assume that this particular problem is more with your code than with how you are structuring this in a machine etc. You need to make your processing async or optimize it a lot.

I was not talking about blocking the JS main thread, but blocking the `currentStateMachine`: the question was about if the `currentStateMachine` would wait until `intervalsMachine` processing the `context` of `currentStatemachine` (the current state) and only then it’ll process the new event (e.g. `GREEN`) and send the `context` (new state) to the `intervalsMachine`.

**Reply by Andarist** [reply]
> I was not talking about blocking the JS main thread, but blocking the currentStateMachine: the question was about if the currentStateMachine would wait until intervalsMachine processing the context of currentStatemachine (the current state) and only then it’ll process the new event (e.g. GREEN) and send the context (new state) to the intervalsMachine.

If we are talking about async processing etc then no - it wont wait automatically and there is no way for you to somehow automatically achieve this behavior right now.

However, that `currentStateMachine` can queue jobs on its own and implement logic based on that - every possible pattern is possible here, you just need to implement it "from scratch". If you could simplify the problem space and list the requirements (I don't even know what "calculate state intervals from the current state" means here, let's assume just some generic data in your requirements) then I could showcase how things could be implemented. It could be then a nice generic example for people, one that wouldn't focus on your business domain.
